File: area_volume.hlm.R
# Version 5 of the hierarchical model for area vs volume
# Colima, and Merapi are unchanneled, whereas shv, unzen and smr are 
# channeled


# Read in data files.
clm = read.table('Clm_Vol_area.txt',header=F);
mrp = read.table('Mrp_Vol_area.txt',header=F);
shv = read.table('Shv_Vol_area.txt',header=F);
unz = read.table('Unz_Vol_area.txt',header=F);
aug = read.table('Aug_Vol_area.txt',header=F);
alldata = list(clm=clm,mrp=mrp,shv=shv,unz=unz,aug=aug);

fnames= c('clm','mrp','shv','unz','aug');
snames = names(alldata);
nvolc = length(alldata);

polys = data.frame(matrix(nrow=2,ncol=5));
colnames(polys) = snames;
njs = rep(0,nvolc);
xj = rep(0,nvolc);
svar = rep(0,nvolc);
for (i in 1:nvolc){
  alldata[[i]][,1] = alldata[[i]][,1]/1000;
  alldata[[i]][,1] = alldata[[i]][,1]^(2/3);
  alldata[[i]] = log10(alldata[[i]]);
  
  linm = lm(alldata[[i]][,2]~alldata[[i]][,1]);
  polys[i] = linm$coefficients;
  njs[i] = length(alldata[[i]][,1]);
  xj[i] = mean(alldata[[i]][,1]);
  svar[i] = sum((alldata[[i]][,1] - xj[i])^2);
}


nsamples = 10000; ########################################### sampling size
nflows = sum(njs);

# Allocating spaces for the dataframes and arrays.
thetas1 = data.frame(matrix(0,nrow=nsamples,ncol=nvolc));
colnames(thetas1) = snames;
thetas2 = data.frame(matrix(0,nrow=nsamples,ncol=nvolc));
colnames(thetas2) = snames;
mus = rep(0,nsamples);
st2.lst = rep(0,nsamples);
sj2.lst = data.frame(matrix(0,nrow=nsamples,ncol=nvolc));
lams = rep(0,nsamples);

# Initializing first values in matrices
t1 = as.numeric(polys[1,]);
t2 = as.numeric(polys[2,]);
thetas1[1,] = t1;
thetas2[1,] = t2;
mus[1] = mean(t2);
sj2init = lapply(alldata,function(x) var(x[,1]));

sj2.lst[1,] = as.numeric(sj2init);
st2.lst[1] = var(t2);

svar1 = svar[1:2];
svar2 = svar[3:nvolc];


vj.func <- function(s.j2,s.t2,Sj.part){(s.j2/Sj.part) + s.t2};
mu.hat.func <- function(t2j,s.j2,s.t2,Sj.part){
  
  sum(t2j/vj.func(s.j2,s.t2,Sj.part))/sum(1/vj.func(s.j2,s.t2,Sj.part))};

mv.rnorm <- function(means,sds){
  n = length(means);
  samples = numeric(length=n);
  for(i in 1:n){
    samples[i] = rnorm(n=1,mean=means[i],sd=sds[i])
  }
  return(samples)
}




q = nsamples/10;
tmp = rep(0,nvolc);

k = 1;
for (ns in 1:(nsamples-1)){
  
  # progress notification
  if (((ns%%q) - 0) < 1e-3){
    #sprintf('%3.1f percent done',ns*10/q);
    #print(ns*10/q); print ('\b% percent done')
    cat(ns*10/q, "% percent done\n")
    flush.console();
  }
  
  ##############################################################
  
  # Step 1: Draw th2
  lamtemp = st2.lst[k] / sj2.lst[k,];
  m2 = t2 - (t2 - mus[k])/(1 + lamtemp * svar);
  s21 = (sj2.lst[k,] * lamtemp)/(1 + lamtemp * svar);
  thetas2[k+1,] = mv.rnorm(as.numeric(m2),as.numeric(sqrt(s21)));
  
  ##############################################################
  
  # Step 2: Draw th1
  m1 = t1 - xj*(thetas2[k+1,] - t2); # use new t2
  s22 = sj2.lst[k,]/njs;
  thetas1[k+1,] = mv.rnorm(as.numeric(m1),as.numeric(sqrt(s22)));
  
  ############################################################## 
  
  # Step 3A: Draw sig2.u
  cu = 0;
  found.u = 0;
  while (found.u != 1){
    btmp = 0;
    for (i in 1:2){
      xji = alldata[[i]][,1];
      yji = alldata[[i]][,2];
      btmp = btmp + sum((yji - (thetas1[k+1,i] + 
                                  thetas2[k+1,i]*xji))^2);
    }
    betau = 0.5 * btmp;
    s2.u = 1/rgamma(n=1,shape=sum(njs[1:2])/2,scale=1/betau);
    
    
    n1 = sum(1/vj.func(as.numeric(sj2.lst[k,]),st2.lst[k],svar)^2);
    b1 = sum(1/vj.func(0,st2.lst[k],svar1)^2);
    b2 = sum(1/vj.func(as.numeric(sj2.lst[k,3:nvolc]),st2.lst[k],svar2)^2);
    
    u.u = runif(1);
    condu = sqrt(n1)/sqrt(b1 + b2);
    if (u.u < condu){
      found.u = 1;
    }
    cu = cu + 1;
  }
  
  ##############################################################
  
  # STEP 3B: Draw sig2.c
  cc = 0;
  found.c = 0;
  
  while (found.c != 1){
    btmp = 0;
    for (i in 3:nvolc){
      xji = alldata[[i]][,1];
      yji = alldata[[i]][,2];
      btmp = btmp + sum((yji - (thetas1[k+1,i] + 
                                  thetas2[k+1,i]*xji))^2);
    }
    betac = 0.5 * btmp;
    s2.c = 1/rgamma(n=1,shape=sum(njs[3:nvolc])/2,scale=1/betac);
    
    n1 = sum(1/vj.func(as.numeric(sj2.lst[k,]),st2.lst[k],svar)^2);
    b1 = sum(1/vj.func(as.numeric(sj2.lst[k,1:2]),st2.lst[k],svar1)^2);
    b2 = sum(1/vj.func(0,st2.lst[k],svar2)^2);
    
    u.c = runif(1);
    condc = sqrt(n1)/sqrt(b1 + b2);
    
    if(u.c < condc){
      found.c = 1;
    }
    cc = cc + 1;
  }
  # Put s2.c and s2.u into the old sj2.lst array
  for (i in 1:2){
    sj2.lst[k+1,i] = s2.u;
  }
  for (i in 3:nvolc){
    sj2.lst[k+1,i] = s2.c;
  }
  
  ##############################################################
  
  # STEP 4: Draw mu
  vs = vj.func(sj2.lst[k+1,], st2.lst[k],svar);
  ms = mu.hat.func(as.numeric(thetas2[k+1,]),
                   as.numeric(sj2.lst[k+1,]),
                   st2.lst[k],svar);
  mus[k+1] = rnorm(n=1,mean=ms,sd=sqrt(1/sum(1/vs)));
  
  ##############################################################
  
  #STEP 5: Generate sig.t2
  beta = 0.5*sum((thetas2[k+1,] - mus[k+1])^2);
  sig.t2.cand = 1/rgamma(n=1,shape=(nvolc-2)/2,scale=1/beta);
  lamcand = sig.t2.cand/sj2.lst[k+1,];
  
  ut = runif(1);
  condt = sqrt(sum(1/vj.func(as.numeric(sj2.lst[k+1,]),sig.t2.cand,svar)^2))/
    sqrt(sum(1/vj.func(as.numeric(sj2.lst[k+1,]),0,svar)^2));
  
  ct = 0;
  while (ut < condt){
    beta = 0.5 * sum((thetas2[k+1,] - mus[k+1])^2);
    sig.t2.cand = 1/rgamma(n=1,shape=(nvolc-2)/2,scale=1/beta);
    lamcand = sig.t2.cand/sj2.lst[k+1,];
    
    ut = runif(1);
    condt = sqrt(sum(1/vj.func(as.numeric(sj2.lst[k+1,]),sig.t2.cand,svar)^2))/
      sqrt(sum(1/vj.func(as.numeric(sj2.lst[k+1,]),0,svar)^2));
    ct = ct + 1;
  }
  st2.lst[k+1] = sig.t2.cand;
  k = k + 1;
}

# number of samples to throw away
burn = 1000;
thetas1 = thetas1[-(1:burn),];
thetas2 = thetas2[-(1:burn),];
mus = mus[-(1:burn)]
st2.lst = st2.lst[-(1:burn)];
sj2.lst = sj2.lst[-(1:burn),];



# save samples
write.table(thetas1,file='results/thetas1_av.Rdata',row.names=F,sep=',')
write.table(thetas2,file='results/thetas2_av.Rdata',row.names=F,sep=',')
write.table(sj2.lst,file='results/sj2_lst_av.Rdata',row.names=F,sep=',')
write.table(mus,file='results/mus_av.Rdata',row.names=F,sep=',')
write.table(st2.lst,file='results/st2_lst_av.Rdata',row.names=F,sep=',')

File: area_volume_hlm.py
# HLM for area vs volume

from numpy import *
import time
import matplotlib.pyplot as plt
from scipy import stats

# load data
clm = recfromtxt('Clm_Vol_area.txt');
mrp = recfromtxt('Mrp_Vol_area.txt');
shv = recfromtxt('Shv_Vol_area.txt');
unz = recfromtxt('Unz_Vol_area.txt');
aug = recfromtxt('Aug_Vol_area.txt');
alldata = [clm,mrp,shv,unz,aug]

name = ['Clm','Mrp', 'SHV','Unz','Aug']
nvolc = len(alldata)
polys = zeros([nvolc,2])
njs = zeros([nvolc,1])
xj = zeros([nvolc,1])
Sj_lst = zeros([nvolc,1])

# Data manipulation: Remember col1 = x, col2 = y
for i in range(nvolc):
    alldata[i][:,0] = alldata[i][:,0]/1000 # change to km^3
    alldata[i][:,0] = alldata[i][:,0]**(2.0/3)
    alldata[i] = log10(alldata[i])
    
    polys[i,:] = polyfit(alldata[i][:,0], alldata[i][:,1],1)
    njs[i] = shape(alldata[i])[0]
    xj[i] = mean(alldata[i][:,0])
    Sj_lst[i] = sum((alldata[i][:,0] - xj[i])**2)

Sj_lst = squeeze(Sj_lst)
njs = squeeze(njs)
xj = squeeze(xj)


# sample size
nsamples = 10**5 #===========================================================
nflows = sum(njs)

# Allocating spaces for arrays
thetas1= zeros([nvolc,nsamples])
thetas2= zeros([nvolc,nsamples])
mus = zeros([nsamples,1])
st2_lst = zeros([nsamples,1])
sj2_lst = zeros([nvolc,nsamples])
lams = zeros([nvolc,nsamples])

# Initializing first values of matrices
t1 = polys[:,1]
t2 = polys[:,0]
thetas1[:,0] = squeeze(t1)
thetas2[:,0] = squeeze(t2)
mus[0] = mean(t2)
sigj2init = map(lambda x: var(x[:,0],ddof=1),alldata)
sj2_lst[:,0] = sigj2init
st2_lst[0] = var(t2,ddof=1)


Sj_lst1 = Sj_lst[:2]
Sj_lst2 = Sj_lst[2:]

vj_func = lambda s_j2,s_t2,Sj_part: (1.0*s_j2/Sj_part) + s_t2

mu_hat_func = lambda t2j, s_j2,s_t2,Sj_part: \
                    sum(t2j/vj_func(s_j2,s_t2,Sj_part))/ \
                    sum(1/vj_func(s_j2,s_t2,Sj_part))

tic = time.time()
q = nsamples/10

foo = zeros([nsamples,2])
for k in range(0,nsamples-1):
    
    # progress report
    if (k%q == 0):
        print "%3.1f %% done"%(k*10/q)
    
    ##########################################################################
    
    # STEP 1: Draw th2
    lamtemp = st2_lst[k]/sj2_lst[:,k]
    m1 = t2 - (t2 - mus[k])/(1 + lamtemp*Sj_lst)
    s21 = (sj2_lst[:,k]*lamtemp)/(1 + lamtemp*Sj_lst)
    thetas2[:,k+1] = squeeze(random.randn(nvolc,1))*sqrt(s21) + m1 

    ##########################################################################
    
    # STEP 2: Draw th1
    m = t1 - xj*(thetas2[:,k+1]-t2) #use new t2
    s22 = sj2_lst[:,k]/njs
    thetas1[:,k+1] = squeeze(random.randn(nvolc,1))*sqrt(s22) + m 

    ##########################################################################

    # STEP 3A: Draw sig2_u
    cu = 0
    found_u = 0
    while (found_u != 1):
        btmp = 0
        for i in range(2):
            xji = alldata[i][:,0]
            yji = alldata[i][:,1]
            btmp = btmp + sum((yji-(thetas1[i,k+1] + xji*thetas2[i,k+1]))**2)

        betau = 0.5*btmp
        s2_u = 1/random.gamma(sum(njs[0:2])/2, 1.0/betau)

        n1 = sum(1/vj_func(sj2_lst[:,k],st2_lst[k],Sj_lst)**2)
        b1 = sum(1/vj_func(0,st2_lst[k],Sj_lst1)**2)
        b2 = sum(1/vj_func(sj2_lst[2:,k],st2_lst[k],Sj_lst2)**2)

        u_u = random.rand()
        condu = sqrt(n1)/sqrt(b1 + b2)

        if (u_u < condu):
            found_u = 1
        cu = cu + 1

    ##########################################################################

    # STEP 3B: Draw sig2_c
    cc = 0
    found_c = 0
    while (found_c != 1):
        btmp = 0
        for i in range(2,nvolc): #i=3:nvolc
            xji = alldata[i][:,0]
            yji = alldata[i][:,1]
            btmp = btmp + sum((yji-(thetas1[i,k+1]+xji*thetas2[i,k+1]))**2)

        betac = 0.5*btmp
        s2_c = 1/random.gamma(sum(njs[2:])/2, 1.0/betac)
        
        n1 = sum(1/vj_func(sj2_lst[:,k],st2_lst[k],Sj_lst)**2)
        b1 = sum(1/vj_func(sj2_lst[:2,k],st2_lst[k],Sj_lst1)**2)
        b2 = sum(1/vj_func(0,st2_lst[k],Sj_lst2)**2)

        u_c = random.rand()
        condc = sqrt(n1)/sqrt(b1 + b2)
        
        if (u_c < condc):
            found_c = 1
        
        cc = cc + 1


    foo[k,0] = cu
    foo[k,1] = cc
    
    for i in range(2):
        sj2_lst[i,k+1] = s2_u
    for i in range(2,5):
        sj2_lst[i,k+1] = s2_c
    
    ##########################################################################
    
    # STEP 4: Draw mu
    vs = vj_func(sj2_lst[:,k+1],st2_lst[k],Sj_lst)
    ms = mu_hat_func(thetas2[:,k+1],sj2_lst[:,k+1],st2_lst[k],Sj_lst)
    mus[k+1] = random.randn(1)*sqrt(1/sum(1/vs)) + ms
    
    ##########################################################################
    
    # STEP 5: Generate sig_t2
    beta = 0.5*sum((thetas2[:,k+1]-mus[k+1])**2)
    sig_t2_cand = 1/random.gamma((nvolc-2)/2, 1.0/beta)
    u_t = random.rand()
    condt = sqrt(sum(1/vj_func(sj2_lst[:,k+1],sig_t2_cand,Sj_lst)**2))/ \
            sqrt(sum(1/vj_func(sj2_lst[:,k+1],0,Sj_lst)**2))

    ct = 0;
    
    while ((u_t < condt) == 0):
        beta = 0.5*sum((thetas2[:,k+1]-mus[k+1])**2)
        sig_t2_cand = 1/random.gamma((nvolc-2)/2, 1.0/beta)

        u_t = random.rand()
        condt = sqrt(sum(1/vj_func(sj2_lst[:,k+1],sig_t2_cand,Sj_lst)**2))/ \
                sqrt(sum(1/vj_func(sj2_lst[:,k+1],0,Sj_lst)**2))
        
        ct = ct + 1
    
    st2_lst[k+1] = sig_t2_cand
    


toc = time.time()
print
print '%10.3f secs elapsed'%(toc-tic)


burn = 1000
thetas1 = thetas1[:,burn:]
thetas2 = thetas2[:,burn:]
mus = mus[burn:]
st2_lst = st2_lst[burn:]
sj2_lst = sj2_lst[:,burn:]



plt.figure(1)
plt.clf()
for i in range(nvolc):
    plt.hist(thetas1[i,:], bins=100, histtype='stepfilled', normed=True,\
            alpha=0.5, label=name[i])
plt.title('Theta1')
plt.legend()



plt.figure(2)
plt.clf()
for i in range(nvolc):
    plt.hist(thetas2[i,:], bins=100, histtype='stepfilled', normed=True,\
            alpha=0.5, label=name[i])
plt.title('Theta2')
plt.legend()



#####################################################################

n = 10;

plt.figure(3)
plt.clf()
figind = 1

for j in range(nvolc):#j = 1 ###################### volcano index

    plt.subplot(3,2,figind)
    figind += 1
    x = alldata[j][:,0]
    y = alldata[j][:,1]
    vv = linspace(-3.2, -1, n);

    p975 = zeros([n,1]);
    p25 = zeros([n,1]);
    p50 = zeros([n,1]);
    fvv = zeros([n,1]);
    
    for i in range(n):
        cf = thetas2[j,:] * vv[i] + thetas1[j,:]
        p25[i] = prctile(cf,p=2.5)
        p50[i] = prctile(cf,p=50)
        p975[i] = prctile(cf,p=97.5)

    nn = len(x)
    p = polyfit(x,y,1)
    sxx = sum((x - mean(x))**2)
    yihat = polyval(p,x)
    seps2 = sum((y - yihat)**2)/(nn-2)
    
    topy = zeros([n,1])
    boty = zeros([n,1])
    
    for i in range(n):
        val = stats.t.ppf(0.975,nn) * sqrt(seps2) * \
                sqrt((1/nn) + ((vv[i] - mean(x))**2/sxx));
        topy[i] = polyval(p,vv[i]) + val;
        boty[i] = polyval(p,vv[i]) - val;
    
    
    

    plt.plot(x,y,'bo',ms=5,label='data')
    plt.plot(vv,p25,'r--', label='HLM - CI')
    plt.plot(vv,p975,'r--')
    plt.plot(vv,p50,'r-',label='HLM - Mean')
    plt.plot(vv,topy, 'k:', label='LR - CI')
    plt.plot(vv,boty, 'k:')
    plt.plot(vv,polyval(p,vv), 'k-',label='LR')
    plt.xlim([-3.2,-1])
    plt.ylim([-2.5,2])
    plt.title(name[j])















